### PR TITLE
More tests for file URL Window drive letter quirk

### PR DESCRIPTION
**Pull request by zcorpan**

### CODE REVIEW

**Comment by Sebmaster** [comment]
There are a few issues with these:
```
  1) Web platform tests parsing </> against <file:///C:/a/b>:

      AssertionError: href
      + expected - actual

      -file:///C:/
      +file:///

      at Context.<anonymous> (test/web-platform.js:20:12)

  2) Web platform tests parsing <//d:> against <file:///C:/a/b>:

      AssertionError: href
      + expected - actual

      -file:///d:
      +file://d:/

      at Context.<anonymous> (test/web-platform.js:20:12)

  3) Web platform tests parsing <//d:/..> against <file:///C:/a/b>:

      AssertionError: href
      + expected - actual

      -file:///d:/
      +file://d:/

      at Context.<anonymous> (test/web-platform.js:20:12)
```

Pretty sure the second and third one are correct failures, not quite sure about the first one.

**Comment by domenic** [comment]
Not sure exactly what you mean by "correct failures" @Sebmaster, but I agree with jsdom/whatwg-url that the first result should be `file:///C:/`. I don't really know what things of the form `//driveletter:` are supposed to do, but I trust jsdom/whatwg-url, so the "actual" results are probably correct. I think at least the spec takes great pains to ensure that there are always three slashes after `file:`

**Comment by Sebmaster** [comment]
"correct failures" as in: I'm pretty sure what whatwg-url does is correct here, mainly because of the triple slash thing (also because Chrome agrees with that resolution too).

**Comment by zcorpan** [comment]
I guess I was confused by reading the spec. But would be nice with confirmation from @annevk that the expected results are actually per spec.

File: url/urltestdata.json [hunk 1]
```diff
@@ -4587,5 +4587,76 @@
     "pathname": "/",
     "search": "",
     "hash": ""
+  },
+  "# More file URL tests by zcorpan",
```
Comment (annevk) [inline url/urltestdata.json]: Why do we put the author here? Should I do that too?

File: url/urltestdata.json [hunk 2]
```diff
@@ -4587,5 +4587,76 @@
...
+    "hostname": "",
+    "port": "",
+    "pathname": "/",
+    "search": "",
+    "hash": ""
+  },
+  {
+    "input": "//d:",
+    "base": "file:///C:/a/b",
+    "href": "file:///d:",
```
Comment (annevk) [inline url/urltestdata.json]: This should end in a slash, no?

**Review by annevk — CHANGES_REQUESTED** [review]

File: url/urltestdata.json (hunk 1 above)
Comment (zcorpan) [inline url/urltestdata.json]: I just wanted to separate these tests from the ones above to direct any blame at me, but of course `git blame` works too. Can remove it though.

File: url/urltestdata.json (hunk 2 above)
Comment (zcorpan) [inline url/urltestdata.json]: https://url.spec.whatwg.org/#file-state "/"
https://url.spec.whatwg.org/#file-slash-state "/"
https://url.spec.whatwg.org/#file-host-state "d:" appended to buffer, then EOF. Buffer is a Windows drive letter.
https://url.spec.whatwg.org/#path-state step 1.4.2 appends buffer to path, without adding a trailing slash.

Comment (annevk) [inline url/urltestdata.json]: That seems correct, but then pathname wouldn't end in a "/" either. It's also not entirely clear to me that is the desired outcome.

Comment (zcorpan) [inline url/urltestdata.json]: https://github.com/whatwg/url/issues/193

File: url/urltestdata.json [hunk 3]
```diff
@@ -4652,7 +4652,7 @@
   {
     "input": "/",
     "base": "file:///C:/a/b",
-    "href": "file:///",
+    "href": "file:///C:/",
     "protocol": "file:",
     "username": "",
```
Comment (domenic) [inline url/urltestdata.json]: pathname should be "/C:/" not "/"

File: url/urltestdata.json [hunk 4]
```diff
@@ -4666,7 +4666,7 @@
   {
     "input": "//d:",
     "base": "file:///C:/a/b",
-    "href": "file://d:/",
+    "href": "file:///d:",
```
Comment (domenic) [inline url/urltestdata.json]: pathname should be "/d:" not "/d:/"

**Review by domenic — APPROVED** [review]
Forgot to fixup some pathnames. Let me push an extra commit...

**Comment by domenic** [comment]
@annevk all looks good to me now. Can we merge this, and deal with https://github.com/whatwg/url/issues/193 as a follow-up?

**Review by annevk — APPROVED** [review]

**Comment by annevk** [comment]
I think that's reasonable. Will let @zcorpan merge in case there was anything else.

**Pull request merged**